vespa-cloud/vespa-documentation-search/src/main/java/ai/vespa/cloud/docsearch/OutLinksDocumentProcessor.java

```diff
@@ -0,0 +1,122 @@
+        public void handleResponse(Response response) {
+            //logger.info("In handleResponse");
+        }
+    }
+
+    @Override
+    public Progress process(Processing processing) {
+        //logger.info("In process");
+        for (DocumentOperation op : processing.getDocumentOperations()) {
+            if (op instanceof DocumentPut) {
+                DocumentPut put = (DocumentPut) op;
+                Document document = put.getDocument();
+                if (document.getDataType().isA(DOC_DOCUMENT_TYPE)) {
+                    String myPath = document.getFieldValue(PATH_FIELD_NAME).toString();
+
+                    Set<String> docsLinkedFromThis = canonicalizeLinks(Path.of(myPath),
+                            onlyFileLinks(removeLinkFragment(document)));
+
+                    Array<StringFieldValue> sfv = new Array<>(DataType.getArray(DataType.STRING));
+                    for (String link : docsLinkedFromThis) {
+                        sfv.add(new StringFieldValue(link));
+                    }
+                    document.setFieldValue(OUTLINKS_FIELD_NAME, sfv);
+
+                    if (docsLinkedFromThis.size() > 0) {
+                        addInLinks(docsLinkedFromThis, myPath, document.getDataType());
+                    }
+                }
+            }
+        }
+        return Progress.DONE;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void addInLinks(Set<String> targetDocs, String myPath, DocumentType docType) {
+        WeightedSet<StringFieldValue> wset = new WeightedSet<>(DataType.getWeightedSet(DataType.STRING));
+        wset.put(new StringFieldValue(myPath), 1);
+        for (String targetDoc : targetDocs) {
+            String targetID = "id:open:" + DOC_DOCUMENT_TYPE + "::open" + targetDoc;
+            DocumentUpdate upd = new DocumentUpdate(docType, targetID);
+            upd.addFieldUpdate(FieldUpdate.createAddAll(docType.getField(INLINKS_FIELD_NAME), wset));
+            Result res = asyncSession.update(upd);
+            if (res.isSuccess()) {
+                //logger.info("Added to requests pending: " + res.getRequestId());
+            } else {
+                //logger.info("Sending Update failed");
+            }
+        }
+    }
+
+    protected static Set<String> onlyFileLinks(Set<String> links) {
+        return links.stream()
+                .filter(l -> !(l.startsWith("http://")))
+                .filter(l -> !(l.startsWith("https://")))
+                .collect(Collectors.toSet());
+    }
+
+    protected static Set<String> canonicalizeLinks(Path docPath, Set<String> links) {
+        // Rewrite links to absolute path
+        // Path is like /documentation/operations/admin-procedures.html
+        return links.stream()
+                .map(l -> docPath.getParent().resolve(l).normalize().toString())
+                .collect(Collectors.toSet());
+    }
+
+    @SuppressWarnings("unchecked")
+    protected static Set<String> removeLinkFragment(Document doc) {
```

Comment:
Chicken and the egg. This should probably scan through the document body and look for links?

Reply:
it gets the already-extracted links in an array of strings. This would be clearer if the method signature was changed to such an array, will consider in next PR yes

Reply:
No, that's fine, but who extracts these, then? It's part of the feed?

Reply:
yes - I'll send a PR shortly for you to merge

Reply:
link extraction in https://github.com/vespa-engine/documentation/pull/1078